tq/basic_download.go

```diff
@@ -42,52 +43,82 @@ func (a *basicDownloadAdapter) WorkerEnding(workerNum int, ctx interface{}) {
 }
 
 func (a *basicDownloadAdapter) DoTransfer(ctx interface{}, t *Transfer, cb ProgressCallback, authOkFunc func()) error {
-	f, fromByte, hashSoFar, err := a.checkResumeDownload(t)
+	// Reserve a temporary filename. We need to make sure nobody operates on the file simultaneously with us.
+	f, err := ioutil.TempFile(a.tempDir(), t.Oid)
 	if err != nil {
 		return err
 	}
-	return a.download(t, cb, authOkFunc, f, fromByte, hashSoFar)
-}
+	defer func() {
+		f.Close()
+		// This will delete temp file if:
+		// - we failed to fully download file and move it to final location including the case when final location already
+		//   exists because other parallel git-lfs processes downloaded file
+		// - we also failed to move it to a partially-downloaded location
+		os.Remove(f.Name())
+	}()
+
+	// Close file because we will attempt to move partially-downloaded one on top of it
+	if err := f.Close(); err != nil {
+		return err
+	}
+
+	// ioutil.TempFile hardcodes 0600 permissions, that's not what we want
+	if err := os.Chmod(f.Name(), 0644); err != nil {
```

Comment:
1. Yes, this is handling a case when there's no resuming and `os.Rename` doesn't rename anything on top of our temporary file.
2. I used `0644` because this is exactly what old code had.
3. Okay, will refactor this to be able to reuse `lfs.TempFile`

Reply:
Force-pushed updated code.